README.md
![alt build:passed](https://travis-ci.org/joseEnrique/parallel-requests.svg?branch=master)

## 1. Install the package
```bash
npm install parallel-requests
```


## 2. Install the command
```bash
npm install parallel-requests -g
```

## Latest release

The version 2.2.0 is the latest stable version of parallel-requests component.
see [release note](https://github.com/joseEnrique/parallel-requests/releases/tag/2.2.0) for details.

For running:

- Download latest version from [2.2.0](https://github.com/joseEnrique/parallel-requests/releases/tag/2.2.0)

CHANGELOG.md
There is not milestone for this version

tests/first-level.test.js
/*!
parallel-requests 2.0.1, built on: 2017-03-15
Copyright (C) 2017 Jose Enrique Ruiz Navarro
http://www.isa.us.es/
https://github.com/joseEnrique/parallel-requests

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.*/


'use strict';

var expect = require('chai').expect;
var lib = require('../src');


/*
 * USE MOCHA AND CHAI for testing your code
 */
describe('First Level test', function() {
    this.timeout(10000);
    it('Execute', (done) => {

        lib.doParallelRequest("https://www.google.es/", "GET", 1).then(function(success) {

            expect(success).to.be.equal(true);


        });



        done();

    });
});


tests/parallel-request.test.js
/*!
parallel-requests 2.2.0, built on: 2017-03-21
Copyright (C) 2017 Jose Enrique Ruiz Navarro
http://www.isa.us.es/
https://github.com/joseEnrique/parallel-requests

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.*/


'use strict';

var expect = require('chai').expect;
var lib = require('../src');


/*
 * USE MOCHA AND CHAI for testing your code
 */
describe('GET google', function() {
    this.timeout(2000);
    it('Execute', (done) => {
        lib.doParallelRequestWithDuration("https://www.google.es/", "GET", 2, 1).then(function(success) {
            expect(success.totalRequest).to.be.equal(2);
            done();
        });
    });
});
describe('POST failed ', function() {
    this.timeout(2000);
    it('Execute', (done) => {
        lib.doParallelRequestWithDuration("https://www.google.es/", "POST", 2, 1).then(function(success) {
            expect(success.error).to.be.equal(2);
            done();
        });
    });
});

describe('POST successful', function() {
    this.timeout(4000);
    it('Execute', (done) => {
        lib.doParallelRequestWithDuration("https://jsonplaceholder.typicode.com/posts", "POST", 2, 1, "{	\"systerminal\": 1}").then(function(success) {
            expect(success.success).to.be.equal(2);
            done();
        });
    });
});


describe('GET only ONE successful', function() {
    this.timeout(4000);
    it('Execute', (done) => {
        lib.doOneStackOfRequest("http://systerminal.com", "GET", 2).then(function(success) {
            console.log(success);
            expect(success.success).to.be.equal(2);
            done();
        });
    });
});



describe('By config file, 3 test successful', function() {
    this.timeout(17000);
    it('Execute', (done) => {
        lib.doParallelRequestFromfile('./tests/config-test.yaml').then(function(success) {
            console.log(success);
            expect(success.length).to.be.equal(3);
            done();

        });

    });
});


command.js
#!/usr/bin/env node

"use strict";
const program = require('commander');
const parallel_requests = require('./index.js')

let bydirectory = (directory) => {
    parallel_requests.doParallelRequestFromfile(directory)
}

let byduration = (data) => {
    //left d
    parallel_requests.doParallelRequestWithDuration(data.url, data.method, data.count, data.duration, data.body);
}

let infinite = (data) => {
    //left d
    parallel_requests.doRequests(data.url, data.method, data.count, data.body);
}




program
    .version('1.1.0')
    .command('readFile')
    .description('read test from file')
    .option('-p, --path <directory>', 'put the filepath')
    .option('-u, --url <url>', 'put the url')
    .action(bydirectory)

program
    .command('timeResquests')
    .description('execute requests from params during any duration')
    .option('-u, --url <url>', 'put the url')
    .option('-c, --count <count>', 'put the count')
    .option('-d, --duration <duration>', 'put the duration')
    .option('-m, --method <method>', 'put the method')
    .option('-b, --body [body]', 'put the body')
    .action(byduration)

program
    .command('infiniteRequests')
    .description('execute requests from params infinitely')
    .option('-u, --url <url>', 'put the url')
    .option('-c, --count <count>', 'put the count')
    .option('-m, --method <method>', 'put the method')
    .option('-b, --body [body]', 'put the body')
    .action(infinite)



program.parse(process.argv);

//node command.js readFile /home/quique/lab/parallel-requests/src/configurations/config-exec.yaml

//node command.js timeResquests -u http://google.es -c 5 -d 2 -m GET
//node command.js infiniteRequests -u http://google.es -c 5  -m GET


src/index.js
/*!
parallel-requests 2.2.0, built on: 2017-03-21
Copyright (C) 2017 Jose Enrique Ruiz Navarro
http://www.isa.us.es/
https://github.com/joseEnrique/parallel-requests

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.*/


'use strict';

var request = require('request');
var Promise = require('bluebird');
var config = require('./configurations/config.js');
var fs = require('fs');
var jsyaml = require('js-yaml');
//var cluster = require('cluster');
var moment = require('moment');


module.exports = {
    doParallelRequestWithDuration: _doParallelRequestWithDuration,
    doParallelRequestFromfile: _doParallelRequestFromfile,
    doOneStackOfRequest: _doOneStackOfRequest,
    doRequests: _doRequests
};



var doRequest = function(options) {
    //console.log("Building promise");
    return new Promise(function(resolve) {
        var success = null;
        request(options, function(error, response) {
            if (!error) {
                if (parseInt(response.statusCode) > 300) {
                    success = false;
                } else {
                    success = true;
                }

                resolve(success);
            } else {
                success = false;
                resolve(success);
            }
        });
    });
};


function _promisesRequests(url, method, number, body) {
    var options;

    if (method.toUpperCase() == "GET") {
        options = {
            uri: url,
            method: method,
        };
    } else {
        if (body) {
            if (typeof(body) == "string") {
                body = JSON.parse(body);
            }
        }
        options = {
            uri: url,
            method: method,
            json: true,
            body: body,
        };

    }

    var asyncTasks = [];

    for (var i = 0; i < number; i++) {
        asyncTasks.push(doRequest(options));
    }
    return Promise.all(asyncTasks);


}


function _doRequests(url, method, count, body) {
    setInterval(function() {
        let total = 0;
        let successful = 0;
        let error = 0;
        let start_date = moment();
        _promisesRequests(url, method, count, body).then(function(success) {
            success.forEach(function(element) {
                if (element) {
                    successful++;
                } else {
                    error++;
                }

                total++;

            });

            let finish_date = moment();
            let resobject = {};
            resobject.totalRequest = total;
            resobject.success = successful;
            resobject.error = error;
            resobject.startTime = start_date.toISOString();
            resobject.finishTime = finish_date.toISOString();
            resobject.lastedFor = finish_date - start_date + " ms";
            console.log(resobject);


        });

    }, 1000); // setInterv0al

}


function _doOneStackOfRequest(url, method, count, body) {
    return new Promise(function(resolve) {
        let total = 0;
        let successful = 0;
        let error = 0;
        let start_date = moment();
        _promisesRequests(url, method, count, body).then(function(success) {
            success.forEach(function(element) {
                if (element) {
                    successful++;
                } else {
                    error++;
                }

                total++;

            });
            let finish_date = moment();
            let resobject = {};
            resobject.totalRequest = total;
            resobject.success = successful;
            resobject.error = error;
            resobject.startTime = start_date.toISOString();
            resobject.finishTime = finish_date.toISOString();
            resobject.lastedFor = finish_date - start_date + " ms";
            console.log(resobject);
            resolve(resobject);

        });


    });


}


function _doParallelRequestWithDuration(url, method, count, duration, body) {

    return new Promise(function(resolve) {
        var promiseRequest = new Promise(function(resolve) {

            let total = 0;
            let successful = 0;
            let error = 0;
            let counter_timer = 0;
            let request_ack = 0;
            let start_date = moment();
            let interval = setInterval(function() {
                if (counter_timer == duration - 1) {
                    clearInterval(interval);
                }

                _promisesRequests(url, method, count, body).then(function(success) {
                    success.forEach(function(element) {
                        if (element) {
                            successful++;
                        } else {
                            error++;
                        }

                        total++;

                    });

                    request_ack++;
                    if (request_ack == duration) {
                        let finish_date = moment();
                        let resobject = {};
                        resobject.totalRequest = total;
                        resobject.success = successful;
                        resobject.error = error;
                        resobject.totalStack = request_ack;
                        resobject.startTime = start_date.toISOString();
                        resobject.finishTime = finish_date.toISOString();
                        resobject.lastedFor = finish_date - start_date + " ms";
                        resolve(resobject);
                    }

                });
                counter_timer++;


            }, 1000); // setInterval

        });

        promiseRequest.then(function(success) {
            success.url = url;
            success.method = method;
            writeLog(success);
            console.log(success);

            resolve(success);



        });
    });
}



function _doParallelRequestFromfile(uri) {
    return new Promise(function(resolve) {
        var configString;
        if (!uri) {
            throw new Error("Parameter URI is required");
        } else {
            configString = fs.readFileSync(uri, "utf8");
        }
        //var newConfigurations = jsyaml.safeLoadAll(configString);
        var newConfigurations = [];

        jsyaml.safeLoadAll(configString, function(doc) {
            newConfigurations.push(doc);
        });

        var configs = newConfigurations.length;
        var executed = 0;
        var arrayResults = [];
        for (var i = 0; i < configs; i++) {
            let singleDoc = newConfigurations[i];
            _doc(singleDoc).then(function(success) {
                executed++;
                arrayResults.push(success);
                if (executed == configs) {
                    resolve(arrayResults);
                }

            });
            //cluster.fork();
        }



    });



}






function _doc(testConfiguration) {
    return new Promise(function(resolve) {
        var type = testConfiguration.type;

        if (type == "interval") {
            let numberOftenants = testConfiguration.tenants.length;
            let executed = 0;
            let arrayOfResult = [];

            for (var element in testConfiguration.tenants) {
                let method = testConfiguration.tenants[element].method ? testConfiguration.tenants[element].method : testConfiguration.method;
                let arrayThrougt = testConfiguration.tenants[element].intervals;
                let duration = testConfiguration.tenants[element].duration;
                let body = testConfiguration.tenants[element].body ? JSON.parse(testConfiguration.tenants[element].body) : undefined;

                execRequestsInterval(testConfiguration.url, method, arrayThrougt, duration, testConfiguration.tenants[element].id, testConfiguration.testId, body).then(function(success) {
                    success.url = testConfiguration.url;
                    success.method = method;
                    success.idTenant = testConfiguration.tenants[element].id;
                    success.idTest = testConfiguration.testId;
                    writeLog(success);
                    arrayOfResult.push(success);
                    executed++;
                    console.log(success);
                    if (executed == numberOftenants) {
                        resolve(arrayOfResult);

                    }

                });

            }

        } else {

            execRequests(testConfiguration).then(function(success) {
                success.url = testConfiguration.url;
                success.method = testConfiguration.method;
                success.idTest = testConfiguration.testId;
                console.log(success);
                writeLog(success);
                resolve(success);


            });


        }

    });

}




var execRequests = function(testConfiguration) {
    return new Promise(function(resolve) {

        let total = 0;
        let successful = 0;
        let error = 0;
        let counter_timer = 0;
        let request_ack = 0;
        let start_date = moment();
        let interval = setInterval(function() {

            if (counter_timer == testConfiguration.duration - 1) {
                clearInterval(interval);
            }

            _promisesRequests(testConfiguration.url, testConfiguration.method, testConfiguration.count, testConfiguration.body).then(function(success) {
                success.forEach(function(element) {
                    if (element) {
                        successful++;
                    } else {
                        error++;
                    }

                    total++;

                });

                request_ack++;
                if (request_ack == testConfiguration.duration) {
                    let finish_date = moment();
                    let resobject = {};
                    resobject.totalRequest = total;
                    resobject.success = successful;
                    resobject.error = error;
                    resobject.totalStack = request_ack;
                    resobject.startTime = start_date.toISOString();
                    resobject.finishTime = finish_date.toISOString();
                    resobject.lastedFor = finish_date - start_date + " ms";
                    resolve(resobject);
                }

            });
            counter_timer++;


        }, 1000); // setInterval

    });

};




var execRequestsInterval = function(uri, method, array, duration, tenantId, testId, body) {
    return new Promise(function(resolve) {
        let total = 0;
        let successful = 0;
        let error = 0;
        let counter_timer = 0;
        let currentindex = 0;
        let request_ack = 0;
        let start_date = moment();
        let interval = setInterval(function() {
            if (counter_timer == duration) {
                //restart counter-time
                counter_timer = 0;
                //logger.info("Next resquests for element in position " + currentindex);
                currentindex++;

                if (currentindex == array.length) {
                    clearInterval(interval);
                }
            }
            _promisesRequests(uri, method, array[currentindex], body).then(function(success) {

                success.forEach(function(element) {
                    if (element) {
                        successful++;
                    } else {
                        error++;
                    }
                    total++;

                });


                if (request_ack == array.length * duration) {
                    let finish_date = moment();
                    let resobject = {};
                    resobject.totalRequest = total;
                    resobject.success = successful;
                    resobject.error = error;
                    resobject.totalStack = request_ack;
                    resobject.startTime = start_date.toISOString();
                    resobject.finishTime = finish_date.toISOString();
                    resobject.lastedFor = finish_date - start_date + " ms";

                    resolve(resobject);
                }

                request_ack++;


            });

            counter_timer++;
        }, 1000);


    });



};

function writeLog(object) {
    fs.appendFile(config.logrequest, JSON.stringify(object) + "\r\n", function(err) {
        if (err) {
            return console.log(err);
        }
    });
}


src/logger/logger.js
/*!
parallel-requests 2.2.0, built on: 2017-03-21
Copyright (C) 2017 Jose Enrique Ruiz Navarro
http://www.isa.us.es/
https://github.com/joseEnrique/parallel-requests

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.*/


'use strict';

/**
 * Module dependecies.
 * */

var winston = require('winston');
var config = require('../configurations/config');

/**
 * Configure here your custom levels.
 * */
var customLeves = {
    levels: {
        error: 7,
        warning: 8,
        custom: 9,
        info: 12,
        debug: 13
    },
    colors: {
        error: 'red',
        warning: 'yellow',
        custom: 'magenta',
        info: 'white',
        debug: 'black'
    }
};

winston.emitErrs = true;

var logger = new winston.Logger({
    levels: customLeves.levels,
    colors: customLeves.colors,
    transports: [
        new winston.transports.File({
            level: config.loglevel,
            filename: config.logfile,
            handleExceptions: true,
            json: false,
            maxsize: 5242880, //5MB
            colorize: false
        }),
        new winston.transports.Console({
            level: config.loglevel,
            handleExceptions: true,
            json: false,
            colorize: true,
            timestamp: true
        })
    ],
    exitOnError: false
});

/*
 * Export functions and Objects
 */
module.exports = logger;

src/configurations/config.js
/*!
parallel-requests 2.2.0, built on: 2017-03-21
Copyright (C) 2017 Jose Enrique Ruiz Navarro
http://www.isa.us.es/
https://github.com/joseEnrique/parallel-requests

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.*/


'use strict';


/**
 * Module dependecies.
 * */

var jsyaml = require('js-yaml');
var fs = require('fs');
var path = require('path');

/*
 * Export functions and Objects
 */
var config = {
    addConfiguration: _addConfiguration
};

module.exports = config;

/*
 * Implement the functions
 */
function _addConfiguration(uri, encoding) {
    var configString = null;

    if (!uri) {
        throw new Error("Parameter URI is required");
    } else {
        configString = fs.readFileSync(path.join(__dirname, uri), encoding);
    }

    var newConfigurations = jsyaml.safeLoad(configString)[process.env.NODE_ENV ? process.env.NODE_ENV : 'development'];

    for (var c in newConfigurations) {
        this[c] = newConfigurations[c];
    }
}

/*
 * Setup default config location
 */
config.addConfiguration('config.yaml', 'utf8');


Gruntfile.js
/*!
parallel-requests 2.2.0, built on: 2017-03-21
Copyright (C) 2017 Jose Enrique Ruiz Navarro
http://www.isa.us.es/
https://github.com/joseEnrique/parallel-requests

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.*/


'use strict';

module.exports = function(grunt) {

    grunt.loadNpmTasks('grunt-contrib-jshint');

    grunt.loadNpmTasks('grunt-contrib-watch');

    grunt.loadNpmTasks('grunt-mocha-test');

    grunt.loadNpmTasks('grunt-release-github');

    grunt.loadNpmTasks('grunt-banner');

    grunt.loadNpmTasks('grunt-dockerize');

    // Project configuration.
    grunt.initConfig({
        //Load configurations
        pkg: grunt.file.readJSON('package.json'),
        licenseNotice: grunt.file.read('extra/license-notice', {
            encoding: 'utf8'
        }).toString(),
        latestReleaseNotes: grunt.file.read('extra/latest-release-notes', {
            encoding: 'utf8'
        }).toString(),

        //Add license notice and latest release notes
        usebanner: {
            license: {
                options: {
                    position: 'top',
                    banner: '/*!\n<%= licenseNotice %>*/\n',
                    replace: true
                },
                files: {
                    src: ['src/**/*.js', 'tests/**/*.js', 'Gruntfile.js'] //If you want to inspect more file, you change this.
                }
            },
            readme: {
                options: {
                    position: 'bottom',
                    banner: '## Latest release\n\n<%= latestReleaseNotes %>',
                    replace: /##\sLatest\srelease(\s||.)+/g,
                    linebreak: false
                },
                files: {
                    src: ['README.md']
                }
            }
        },

        //Lint JS
        jshint: {
            all: ['Gruntfile.js', 'src/**/*.js', 'tests/**/*.js'], //If you want to inspect more file, you change this.
            options: {
                jshintrc: '.jshintrc'
            }
        },

        //Execute mocha tests
        mochaTest: {
            tests: {
                options: {
                    reporter: 'spec',
                    //captureFile: 'test.results<%= grunt.template.today("yyyy-mm-dd:HH:mm:ss") %>.txt', // Optionally capture the reporter output to a file
                    quiet: false, // Optionally suppress output to standard out (defaults to false)
                    clearRequireCache: false, // Optionally clear the require cache before running tests (defaults to false)
                    noFail: false // Optionally set to not fail on failed tests (will still fail on other errors)
                },
                src: ['tests/**/*.js']
            }
        },

        //Make a new release on github
        //"grunt release" for pacth version
        //"grunt release:minior" for minior version
        //"grunt release:major" for major version
        release: {
            options: {
                changelog: true, //NOT CHANGE
                changelogFromGithub: true, //NOT CHANGE
                githubReleaseBody: 'See [CHANGELOG.md](./CHANGELOG.md) for details.', //NOT CHANGE
                npm: false, //CHANGE TO TRUE IF YOUR PROJECT IS A NPM MODULE
                //npmtag: true, //default: no tag
                beforeBump: [], // IS NOT READY YET
                afterBump: [], // IS NOT READY YET
                beforeRelease: [], // IS NOT READY YET
                afterRelease: [], // IS NOT READY YET
                updateVars: ['pkg'], //NOT CHANGE
                github: {
                    repo: "joseEnrique/parallel-requests",
                    accessTokenVar: "GITHUB_ACCESS_TOKEN", //SET ENVIRONMENT VARIABLE WITH THIS NAME
                    usernameVar: "GITHUB_USERNAME" //SET ENVIRONMENT VARIABLE WITH THIS NAME
                }
            }
        },

        //IT IS RECOMENDED TO EXECUTE "grunt watch" while you are working.
        watch: {
            scripts: {
                files: ['src/**/*.js'],
                tasks: ['jshint']
            }
        },

        /*
                dockerize: {
                    'project-template-latest': {
                        options: {
                            auth: {
                                email: "DOCKER_HUB_EMAIL",
                                username: "DOCKER_HUB_USERNAME",
                                password: "DOCKER_HUB_PASSWORD"
                            },
                            name: 'parallel-requests',
                            push: true
                        }
                    },
                    'project-template-version': {
                        options: {
                            auth: {
                                email: "DOCKER_HUB_EMAIL",
                                username: "DOCKER_HUB_USERNAME",
                                password: "DOCKER_HUB_PASSWORD"
                            },
                            name: 'parallel-requests',
                            tag: '<%= pkg.version %>',
                            push: true
                        }
                    }
                }
                */
    });

    // Default task(s).
    grunt.registerTask('default', ['jshint', 'usebanner']);

    //TEST TASK
    grunt.registerTask('test', ['jshint', 'mochaTest']);

    //BUILD TASK
    grunt.registerTask('build', ['test', 'usebanner']);

    //DEVELOPMENT TASK
    grunt.registerTask('dev', ['watch']);

};
